core/trino-main/src/main/java/io/trino/util/FastutilSetHelper.java

```diff
@@ -203,16 +229,93 @@ public boolean equals(Object a, Object b)
+                        () -> new GeneratedMethod<>(type, operatorInterface, methodHandle));
+                return generatedMethod.get();
+            }
+            catch (ExecutionException | UncheckedExecutionException e) {
+                throwIfUnchecked(e.getCause());
+                throw new RuntimeException(e.getCause());
             }
-            catch (Throwable t) {
-                throwIfInstanceOf(t, Error.class);
-                throwIfInstanceOf(t, TrinoException.class);
-                throw new TrinoException(GENERIC_INTERNAL_ERROR, t);
+        }
+
+        private static class MethodKey<T>
+        {
+            private final Type type;
+            private final Class<T> singleMethodInterface;
+
+            public MethodKey(Type type, Class<T> singleMethodInterface)
+            {
+                this.type = requireNonNull(type, "type is null");
+                this.singleMethodInterface = requireNonNull(singleMethodInterface, "singleMethodInterface is null");
+            }
+
+            @Override
+            public boolean equals(Object o)
+            {
+                if (this == o) {
+                    return true;
+                }
+                if (o == null || getClass() != o.getClass()) {
+                    return false;
+                }
+                MethodKey<?> that = (MethodKey<?>) o;
+                return type.equals(that.type) && singleMethodInterface.equals(that.singleMethodInterface);
+            }
+
+            @Override
+            public int hashCode()
+            {
+                return Objects.hash(type, singleMethodInterface);
+            }
+        }
+
+        private static class GeneratedMethod<T>
+        {
+            private Type type;
+            private Class<T> singleMethodInterface;
+            private MethodHandle methodHandle;
+            @GuardedBy("this")
+            private T operator;
+
+            public GeneratedMethod(Type type, Class<T> singleMethodInterface, MethodHandle methodHandle)
+            {
+                this.type = requireNonNull(type, "type is null");
+                this.singleMethodInterface = requireNonNull(singleMethodInterface, "singleMethodInterface is null");
+                this.methodHandle = requireNonNull(methodHandle, "methodHandle is null");
+            }
+
+            public synchronized T get()
+            {
+                if (operator != null) {
+                    return operator;
+                }
+                String suggestedClassName = singleMethodInterface.getSimpleName() + "_" + type.getDisplayName();
+                operator = compileSingleAccessMethod(suggestedClassName, singleMethodInterface, methodHandle);
+                // Free up memory by removing references to the objects that we won't need anymore
+                type = null;
+                singleMethodInterface = null;
+                methodHandle = null;
+                return operator;
```

Comment:
Actually a question. Is there a reason to do this part lazyily. Why not eagerly generate `T` and store it in cache?

Reply:
For this I was trying to stay close to `BlockTypeOperators#getBlockOperator` as that is doing very similar thing. I don't see any drawback in the simpler eager generation approach though. Either way we should end up holding lock on only one entry.
@dain please chime in if there is a specific reason to prefer the lazy loading from `BlockTypeOperators#getBlockOperator`.

Reply:
I tried to dig a bit into the granularity of locking used by guava Cache. I get the impression that it locks a segment of the Map only to insert entry (which we would encounter either way), but for the load of the entry it takes lock only on the newly inserted entry which should be be similar to what we did here.
But I'm happy to stick to the current way for the sake of consistency and simpler exception handling.

Reply:
IIRC Guava uses striped lockes.  Typically a system like this uses 1 lock per core.  I think the bigger win is a simpler stack when things go wrong.  Also, in the case of a failure, you get one total object instead of one object per request to the cache